src/eos-prepare-usb-update.c

```diff
@@ -0,0 +1,665 @@
+  g_autoptr(GVariant) options = NULL;
+  g_autoptr(GError) local_error = NULL;
+  const gchar *refs[] = { pull_data->refspec->ref };
+  const gchar *override_commit_ids[] = { pull_data->commit_id };
+  g_autoptr(ScopedMainContext) context = scoped_main_context_new ();
+
+  g_variant_builder_init (&builder, G_VARIANT_TYPE_VARDICT);
+  g_variant_builder_add (&builder,
+                         "{s@v}",
+                         "override-url",
+                         g_variant_new_variant (g_variant_new_string (pull_data->server_uri)));
+  g_variant_builder_add (&builder,
+                         "{s@v}",
+                         "refs",
+                         g_variant_new_variant (g_variant_new_strv (refs, 1)));
+  g_variant_builder_add (&builder,
+                         "{s@v}",
+                         "override-commit-ids",
+                         g_variant_new_variant (g_variant_new_strv (override_commit_ids, 1)));
+  g_variant_builder_add (&builder,
+                         "{s@v}",
+                         "depth",
+                         g_variant_new_variant (g_variant_new_int32 (0)));
+  options = g_variant_ref_sink (g_variant_builder_end (&builder));
+
+  if (!ostree_repo_pull_with_options (repo,
+                                      pull_data->refspec->remote,
+                                      options,
+                                      pull_data->progress,
+                                      cancellable,
+                                      &local_error))
+    {
+      g_task_return_error (task, g_steal_pointer (&local_error));
+      return;
+    }
+
+  g_task_return_boolean (task, TRUE);
+}
+
+static void
+run_ostree_pull_in_thread (OstreeRepo *repo,
+                           GCancellable *cancellable,
+                           EosPullData *pull_data)
+{
+  g_autoptr(GTask) task = g_task_new (repo,
+                                      cancellable,
+                                      pull_ready,
+                                      g_object_ref (pull_data));
+
+  g_task_set_task_data (task, g_object_ref (pull_data), g_object_unref);
+  g_task_run_in_thread (task, run_pull_task_func);
+}
+
+static gboolean
+do_pull (OstreeRepo *source_repo,
+         OstreeRepo *target_repo,
+         EosRefspec *refspec,
+         const gchar *commit_id,
+         OstreeAsyncProgress *progress,
+         GCancellable *cancellable,
+         GError **error)
+{
+  g_autoptr(ScopedMainContext) context = scoped_main_context_new ();
+  g_autoptr(GMainLoop) loop = NULL;
+  g_autoptr(EosUpdaterRepoServer) server = NULL;
+  g_autofree gchar *server_uri = NULL;
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(EosPullData) pull_data = NULL;
+
+  if (!get_server_and_uri (source_repo,
+                           refspec,
+                           cancellable,
+                           &server,
+                           &server_uri,
+                           error))
+    return FALSE;
```

Comment:
What's the purpose of spinning up a server to pull from? We can just pass in the path to the source repo and do the equivalent of `pull-local`, which will avoid all the HTTP overhead. Oh, I guess this is because the server will be a fake archive-z2 repo. But if the USB preparation is running at root, the pull can be done directly. OSTree handles this fine as long as the source repo uri scheme is `file`.